s02/2.md

```diff
@@ -10,14 +10,556 @@ Tartalom
+* mindig publikus
+* nincs típusa
+* nincs paramétere
+* nem túlterhelhető
+
+```C++
+class <típus> {
+    public:
+    <typusnév>() { … } // konstruktor
+    ~<típusnév>() { … } // destruktor
+    …
+};
+```
+
+### Objektumok másolása
+
+Kétféle másolási megközelítés ismert:
+
+* _Sekély másolás (shallow copy)_: A típuspéldány a mezőivel együtt másolásra kerül egy új memóriaterületre.
+    A dinamikusan lefoglalt mezőknek azonban az értéke nem másolódik (A régi és új példány mutatói ugyanazon területre
+    fognak mutatni.)
+
+* _Mély másolás (deep copy)_: A típuspéldány minden mezőjével és azok által lefogalt memóriaterülettel együtt
+    kerül másolásra. (A régi és az új példány mutatói nem ugyanazon területre fognak mutatni.)
+
+Példányok másolását két művelet teszi lehetővé: _másoló konstruktor_, _értékadó operátor_
+
+#### Copy-konstruktor
+
+Egy létező példány alapján újat hoz létre. Paraméterként egy másik (ugyanolyan típusú) példány referenciáját kapja,
+ennek a mezőit másolja le. (Ha nincs dinamikus tartalom, akkor az alapértelmezett megfelelő.)
+
+* törzsben tud hivatkozni a másolandó példány mezőire
+* a következő esetekben fut le:
+    + közvetlen hívás: `MyType b(a);`
+    + kezdeti értékadás: `MyType b = a;`
+    + érték szerinti paraméterátadás
+
+```C++
+class MyType {
+    private:
+        int* _value;
+    public:
+        MyType(const MyType& other) { // másoló konstr.
+            _value = new int;
+
+            // a dinamikus taralom létrehozása
+            *_value = *other._value; // érték másolása
+        }
+};
+```
+
+#### Értékadó operátor
+
+A kezdeti értékadást kivéve, amikor a változónak értéket adunk, az értékadó operátor lép érvénybe.
+
+Megkapja a másolandó példány (konstans) referenciáját, és biztosítja taralmának átmásolását.
+
+* az eddig meglévő, dinamikusan létrehozott értékeket törölni kell
+* ellenőrizni kell, hogy a paraméterben kapott változó nem saját maga-e
+* a `*this` (aktuális példány) referenciával kell visszatérni (a többszörös értékadás használatához)
+
+```C++
+class MyType {
+    public:
+    …
+        MyType& operator=(const MyType& other){
+            if (this == &other)
+                // ha ugyanazt a példányt kaptuk
+                return *this; // nem csinálunk semmit
+
+            *_value = *(other._value);
+            // különben a megfelelő módon másolunk
+            return *this; // visszaadjuk a referenciát
+        }
+};
+```
+
+#### Paraméterátadás
+
+Ahogy a változókat, úgy a paramétereket is háromféleképpen tudjuk átadni: `érték szerint`, `referenciaként`, `pointerként`
+
+Az érték szerinti átadás sokszor költséges lehet, mert ekkor a paraméterek másolódnak. A pointer és a referencia közötti
+döntést pedig az határozza meg, hogy míg a pointerek felvehetik a `NULL` értéket, addig a referenciák nem.
+
+Tehát a következők szerint érdemes a paraméterátadást használni:
+
+* _érték szerinti_:
+    + a függvény **nem módosítja a paramétert**
+    + **könnyű másolni**
+    
```

Comment:
"2. Pass by const pointer when the value is expensive to copy AND the function does
not want to modify the value pointed to AND NULL is a valid, expected value that
the function handles."

Itt nem azt mondja, hogy "költséges a másolás", hanem hogy "használd ezt, ha az érték másolása költséges lenne".

Most látom én is félrenéztem, ezek a "költséges másolni" és "könnyű másolni" dolgok bezavartak nekem. Az egészet átírtam most:

* _érték szerinti_:
    + ```f(int x)```
    + a függvény **nem módosíthatja a paramétert**
    + használd ezt **ha könnyű másolni**
    + Ökölszabály: ha a paraméter mérete legfeljebb 2- vagy 3 szó (_word_), azaz 32-bit, akkor érdemes érték szerint átadni
        - egyszerű primitív típusok esetén ajánlott, mint pl.: `int`, `double`, `char`, `bool`, stb.
        - Komplex típusok, saját osztályok, `std::string` és a különböző STL konténerek nem ajánlottak (továbbiakban ezeket a példákban `T`-vel jelöljük)
* _pointer szerinti_:
    + ```f(T * x)```
    + a függvény **módosíthatja a paramétert**,
    + használd ezt, **ha költséges a másolás**,
    + továbbá akkor, ha **NULL lehet valid** érték
* _konstans pointer szerinti_:
    + ```f(const * T x)```
    + a függvény **módosíthatja az értéket amire mutat, de nem magát a pointert**,
    + használd ezt, **ha költséges a másolás**, 
    + továbbá akkor, ha **NULL lehet valid** érték
* _referencia szerinti_: 
    + ```f(T& x)```
    + a függvény **módosíthatja a paramétert**
    + használd ezt, **ha költséges a másolás**,
    + továbbá a **NULL NEM lehet valid** érték
    + használható kiementként is
* _konstans referencia szerinti_:
    + ```f(const T& x)```
    + a függvény **nem módosíthatja a paramétert**,
    + használd ezt, **ha költséges a másolás**
    + továbbá ha **NULL NEM lehet valid** érték
    + próbáljuk mindig ezt a változatot használni komplex típusokhoz, saját osztályokhoz, `std::string`-hez és STL adatszerkezetekhez a biztonságosabb programozás érdekében.
        - lásd még: copy konstruktorok, copy assignment-ek paraméterei